libraries/reconstruction_cuda/cuda_gpu_iirconvolve.cu

```diff
@@ -1,107 +1,314 @@
-    }
-    line[(xDim - 1) * yDim] *= z / (z - 1.f);
-    for (int j = xDim - 2; 0 <= j; --j) {
-        line[j * yDim] = z * (line[(j + 1) * yDim] - line[j * yDim]);
+__global__ void rows_iterate(data_ptr in, int x) {
+    __shared__ float sdata[BLOCK_SIZE][BLOCK_SIZE + 1];
+
+    forward_pass(in, sdata, x);
+    backward_pass(in, sdata, x);
+}
+
+
+__global__ void cols_iterate_switch(data_ptr in, data_ptr out, int x, int y) {
+    const float z = sqrtf(3.f) - 2.f;
+    const float k = z / (z - 1.f);
+
+    const float gain = 6.0f;
+
+    int col = blockIdx.x * blockDim.x + threadIdx.x;
+
+    if (col < x) {
+        float *myCol = in + col;
+        float *outCol = out + col;
+        float sum = myCol[0];
+
+        outCol[0] = sum * z * gain;
+        #pragma unroll UNROLL
+        for (int j = 1; j < y; ++j) {
+            outCol[j*x] = myCol[j*x] * gain + z * outCol[(j - 1)*x];
+        }
+
+        outCol[(y - 1)*x] *= k;
+        #pragma unroll UNROLL
+        for (int j = y - 2; 0 <= j; --j) {
+            outCol[j*x] = z * (outCol[(j + 1)*x] - outCol[j*x]);
+        }
     }
+
 }
 
+__global__ void cols_iterate(data_ptr in, int x, int y) {
+    const float z = sqrtf(3.f) - 2.f;
+    const float k = z / (z - 1.f);
+    const float gain = 6.0f;
 
-__global__
-void iirConvolve2D_Cardinal_Bspline_3_MirrorOffBound(float* input,
-        float* output, size_t xDim, size_t yDim) {
-    // assign line to thread
-    int idy = blockIdx.y * blockDim.y + threadIdx.y;
-    if (idy >= yDim) return; // only threads with data should continue
-    float* line = output + (idy * xDim);
-
-    // adjust gain
-    float z = sqrtf(3.f) - 2.f;
-    float z1 = 1.0 - z;
-    float gain = -(z1 * z1) / z;
-
-    // copy original data
-    for (int i = 0; i < xDim; i++) {
-        line[i] = input[i + (idy * xDim)] * gain;
-    }
-    // prepare some values
-    float sum = (line[0] + powf(z, xDim) * line[xDim - 1]) * (1.f + z) / z;
-    z1 = z;
-    float z2 = powf(z, 2 * xDim - 2);
-    float iz = 1.f / z;
-    for (int j = 1; j < (xDim - 1); ++j) {
-        sum += (z2 + z1) * line[j];
-        z1 *= z;
-        z2 *= iz;
-    }
-    line[0] = sum * z / (1.f - powf(z, 2 * xDim));
-    for (int j = 1; j < xDim; ++j) {
-        line[j] += z * line[j - 1];
-    }
-    line[xDim - 1] *= z / (z - 1.f);
-    for (int j = xDim - 2; 0 <= j; --j) {
-        line[j] = z * (line[j + 1] - line[j]);
+    const int col = blockIdx.x * blockDim.x + threadIdx.x;
+    float *myCol = in + col;
+
+    if (col < x) {
+
+        myCol[0] *= z * gain;
+        #pragma unroll UNROLL
+        for (int j = 1; j < y; ++j) {
+            myCol[j*x] = myCol[j*x] * gain + z * myCol[(j - 1)*x];
+        }
+
+        myCol[(y - 1)*x] *= k;
+        #pragma unroll UNROLL
+        for (int j = y - 2; 0 <= j; --j) {
+            myCol[j*x] = z * (myCol[(j + 1)*x] - myCol[j*x]);
+        }
     }
 }
+
+/*
+ * Faster when compiled with fixed `x` and `y`
```

Comment:
Maybe we can try to identify the most commonly used sizes (@cossorzano) and specialize the kernels?